// parrot-core/lib/utils/logger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loggerColors = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var loggerColors = exports.loggerColors = {
  info: _chalk2.default.white,
  warn: _chalk2.default.yellow,
  error: _chalk2.default.red
}; /* eslint no-console: 'off' */


var infoTemplate = function infoTemplate(message) {
  return loggerColors.info('Info: ' + message);
};
var errorTemplate = function errorTemplate(message) {
  return loggerColors.error('Error: ' + message);
};
var warnTemplate = function warnTemplate(message) {
  return loggerColors.warn('Warning: ' + message);
};

var Logger = function () {
  function Logger() {
    var output = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : console.log;
    (0, _classCallCheck3.default)(this, Logger);

    this.output = output;
    this.info = this.baseTemplate(infoTemplate);
    this.error = this.baseTemplate(errorTemplate);
    this.warn = this.baseTemplate(warnTemplate);
  }

  (0, _createClass3.default)(Logger, [{
    key: 'baseTemplate',
    value: function baseTemplate(template) {
      var _this = this;

      return function (message, path) {
        return _this.output('[Parrot] ' + _chalk2.default.underline(path) + (' ' + _chalk2.default.dim('(' + _this.scenario + ')') + '\n\t' + template(message)));
      };
    }
  }, {
    key: 'setScenario',
    value: function setScenario(scenario) {
      this.scenario = scenario;
      return this;
    }
  }]);
  return Logger;
}();

exports.default = new Logger();
